src/ziggurat/kafka_consumer/consumer_handler.clj

```diff
@@ -0,0 +1,84 @@
+(ns ziggurat.kafka-consumer.consumer-handler
+  (:require [clojure.tools.logging :as log]
+            [ziggurat.config :refer :all]
+            [ziggurat.messaging.producer :as producer]
+            [ziggurat.message-payload :refer [map->MessagePayload]]
+            [ziggurat.metrics :as metrics])
+  (:import (org.apache.kafka.common.errors WakeupException)
+           (java.time Duration Instant)
+           (org.apache.kafka.clients.consumer Consumer ConsumerRecord)))
+
+(def batch-consumption-metric-ns ["ziggurat.batch.consumption" "message.processed"])
+
+(defn- publish-batch-process-metrics
+  [topic-entity batch-size success-count skip-count retry-count time-taken-in-millis]
+  (let [topic-entity-tag {:topic-entity topic-entity}]
+    (metrics/increment-count batch-consumption-metric-ns "total" batch-size topic-entity-tag)
+    (metrics/increment-count batch-consumption-metric-ns "success" success-count topic-entity-tag)
+    (metrics/increment-count batch-consumption-metric-ns "skip" skip-count topic-entity-tag)
+    (metrics/increment-count batch-consumption-metric-ns "retry" retry-count topic-entity-tag)
+    (metrics/report-time (conj batch-consumption-metric-ns "execution-time") time-taken-in-millis topic-entity-tag)))
+
+(defn- retry
+  ([batch-payload]
+   (producer/retry batch-payload))
+  ([batch current-retry-count topic-entity]
+   (when (pos? (count batch))
+     (let [message (map->MessagePayload {:message         batch
+                                         :retry-count        current-retry-count
+                                         :topic-entity topic-entity})]
+       (producer/retry message)))))
+
+(defn process
+  [batch-handler batch-payload]
+  (let [batch               (:message batch-payload)
+        topic-entity        (:topic-entity batch-payload)
+        current-retry-count (:retry-count batch-payload)
+        batch-size          (count batch)]
+    (try
+      (when (not-empty batch)
+        (log/infof "[Consumer Group: %s] Processing the batch with %d messages" topic-entity batch-size)
+        (let [start-time             (Instant/now)
+              result                 (batch-handler batch)
+              time-taken-in-millis   (.toMillis (Duration/between start-time (Instant/now)))]
+          (when (map? result)
+            (let [messages-to-be-retried (:retry result)
```

Comment:
might be better if you merge the 2 when expressions into one

Reply:
Between these 2 when statement we are processing the batch and measuring the time taken to process the batch. So can't merge 2 when expressions as both are used for different purpose.

Reply:
you can check the `batch` is `not-empty` and the `result` is a `map?` on the first lest expression so you can have just 1 when expression